src/ServiceControl/Notifications/Email/SendEmailNotificationHandler.cs

```diff
@@ -0,0 +1,115 @@
+﻿namespace ServiceControl.Notifications.Email
+{
+    using System;
+    using System.Threading.Tasks;
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using NServiceBus.Transport;
+    using Raven.Client;
+    using ServiceBus.Management.Infrastructure.Settings;
+
+    public class SendEmailNotificationHandler : IHandleMessages<SendEmailNotification>
+    {
+        readonly IDocumentStore store;
+        readonly EmailThrottlingState throttlingState;
+
+        public SendEmailNotificationHandler(IDocumentStore store, Settings settings, EmailThrottlingState throttlingState)
+        {
+            this.store = store;
+            this.throttlingState = throttlingState;
+
+            emailDropFolder = settings.EmailDropFolder;
+        }
+
+        public async Task Handle(SendEmailNotification message, IMessageHandlerContext context)
+        {
+            NotificationsSettings notifications;
+
+            using (var session = store.OpenAsyncSession())
+            {
+                notifications = await session.LoadAsync<NotificationsSettings>(NotificationsSettings.SingleDocumentId)
+                    .ConfigureAwait(false);
+            }
+
+            if (notifications == null || !notifications.Email.Enabled)
+            {
+                log.Info("Skipping email sending. Notifications turned-off.");
+                return;
+            }
+
+            var hasSemaphore = false;
+
+            try
+            {
+                while (hasSemaphore == false)
+                {
+                    if (throttlingState.IsThrottling())
+                    {
+                        log.Warn("Email notification throttled");
+                        return;
+                    }
+
+                    hasSemaphore = await throttlingState.Semaphore.WaitAsync(spinDelay).ConfigureAwait(false);
+                }
+
+                if (context.MessageId == throttlingState.RetriedMessageId)
+                {
+                    message.Body +=
+                        "\n\nWARNING: Your SMTP server was temporarily unavailable. Make sure to check ServicePulse for a full list of health check notifications.";
+                }
+
+                await EmailSender.Send(notifications.Email, message.Subject, message.Body, emailDropFolder)
+                    .ConfigureAwait(false);
+            }
+            catch (Exception e) when (!(e is OperationCanceledException))
+            {
+                if (message.IsFailure)
+                {
+                    throttlingState.ThrottlingOn();
+
+                    await Task.Delay(throttlingDelay).ConfigureAwait(false);
+
+                    throttlingState.ThrottlingOff();
+
+                    throttlingState.RetriedMessageId = context.MessageId;
+
+                    throw new EmailNotificationException(e);
```

Comment:
If I am reading this right, when an exception occurs, we turn throttling on which causes all pending operations to simply stop processing, consuming their message. Any new requests that come in during that time will also get dropped without any attempt to send an email. After a while we turn throttling off but then throw an exception, forcing this message to be retried. Why is this email notification more important than all of the ones we dropped? Why not just drop this one?

Reply:
First, thanks for a great write-up on what is happening in terms of throttling.

The reason for this behavior is the following. We want to make sure that a user gets at least one failure notification even if the SMTP is down - that's why we throw `EmailNotificationException` and always do immediate retires for this exception. At the same we want to prevent messages from piling up in ServiceControl input queue - that's why we drop all messages, except one for which an exception was thrown, when in throttling mode.

The main UX argument for this behavior is that all the notifications are important enough that the users are very likely to look into ServicePulse when they any. At that point, they will see all the other failures for which we potentially dropped the notification. To make it a bit more visible to the user email body gets an additional line that prompts to look into ServicePulse for other notifications when there were some SMTP connectivity issues..